tests/unit/s2n_server_hello_test.c

```diff
@@ -173,6 +173,155 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_disable_tls13());
     }
 
+    /* Test TLS 1.3 => 1.1 protocol downgrade detection with a TLS1.3 client */
+    {
+        EXPECT_SUCCESS(s2n_enable_tls13());
+        struct s2n_config *client_config;
+        struct s2n_connection *client_conn;
+        struct s2n_config *server_config;
+        struct s2n_connection *server_conn;
+
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        EXPECT_NOT_NULL(client_config = s2n_config_new());
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+
+        /* The client will request TLS1.3 */
+        client_conn->client_protocol_version = S2N_TLS13;
+
+        struct s2n_stuffer *server_stuffer = &server_conn->handshake.io;
+
+        const uint32_t total = S2N_TLS_PROTOCOL_VERSION_LEN
+            + S2N_TLS_RANDOM_DATA_LEN
+            + SESSION_ID_SIZE
+            + server_conn->session_id_len
+            + S2N_TLS_CIPHER_SUITE_LEN
+            + COMPRESSION_METHOD_SIZE;
+
+        /* The server will respond with TLS1.1 even though it supports TLS1.3 */
+        server_conn->actual_protocol_version = S2N_TLS11;
+        EXPECT_SUCCESS(s2n_server_hello_send(server_conn));
```

Comment:
for posterity, perhaps check that we have the last downgrade protect bytes?

Reply:
I don't follow. On line 214 we check that we received the downgrade error. What should we be looking for here?

Reply:
I would just verify the bytes to test s2n_server_hello_recv() did set the downgrade bytes. s2n_server_hello_recv() kinds of does it, it's something I prefer to be more explicit so the tests doesn't depend on the implementation of s2n_server_hello_recv() to test that.

Reply:
The implementation of s2n_server_hello_recv() is what we're testing here. If the implementation changes, we want to fail. I can come up with contrived situations where somebody modifies the conditionals in a way where this test passes, but the bytes aren't actually set.  Regardless, I've added this check for each of the test cases.